File: 0046-permutations/0046-permutations.cpp
class Solution {
    vector<vector<int>> ans;
    void dfs(int start,int size,vector<int>& nums){
        if(start>=size){
            ans.push_back(nums);
            return;
        }
        else{
            for(int i=start;i<size;++i){
                swap(nums[start],nums[i]);
                dfs(start+1,size,nums);
                swap(nums[start],nums[i]);
            }
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        dfs(0,nums.size(),nums);
        return ans;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        recurPermute(0,nums,ans);
        return ans;
    }
    private void recurPermute(int idx, int[] nums, List<List<Integer>> ans){
        if(idx == nums.length){
            List<Integer> ds = new ArrayList<>();
            for(int num : nums){
                ds.add(num);
            }
            ans.add(ds);
            return;
        }

        for(int i=idx; i<nums.length; i++){
            swap(i,idx,nums);
            recurPermute(idx+1,nums,ans);
            swap(i,idx,nums);
        }
    }
    private void swap(int i,int idx,int[] nums){
        int temp = nums[i];
        nums[i] = nums[idx];
        nums[idx] = temp;
    }
}